### ISSUE TITLE
cursor: mousebindings refactor + check for mouse button leaks when used in Frame context

### ISSUE THREAD

**Issue by Consolatis** [labwc/labwc#624]
Copied over from https://github.com/labwc/labwc/pull/610#issuecomment-1306384695

The `switch()` within the loop doing `break` and `continue` is not exactly easy to grasp though (`break` corresponding to the `switch()` while `continue` corresponds to the outer loop), this might deserve another Issue / PR.

I wonder though.. don't we leak stray press / release events?
E.g. when an action is bound to `Frame` + `Press` we would leak the corresponding release event (if there is no `Frame` handler for that as well). And similar when an action is bound to `Frame` + `Release` we would leak the corresponding press event.

That means that we can't actually properly implement `Release` mousebindings for `Frame` without also leaking the press event. And we should set `mousebind->pressed_in_context` when we execute a `Frame` + `Press` mousebind which gets cleared on the release handler (+ consumes the release event).

`Click` bindings are handled properly though because we mark them as pressed in the `Press` handler and execute them later within the `Release` handler. `Drag` bindings look like they leak a stray release event as well because we reset `mousebind->pressed_in_context` in `process_cursor_motion()`. I didn't test any of this though, probably something for a separate issue. 

So to sum up:
- The `switch()` within the loop using `break` and `continue` is hard to grasp IMO and would benefit from a refactor
- I think we are leaking some mouse events for the `Frame` context to the application.

**Comment by bi4k8** [comment]
Anecdotally, I have seen applications complain about unpaired release events.

**Comment by Tamaranch** [comment]
Not sure if my comment is in the right place or if it deserves a new issue. Let's try here, it's a corner case anyway.

When a surface is clicked and is hidden before the button is released, the corresponding release event is never received (neither when the surface is hidden nor when the button is released). On Wayfire it is received when the surface is hidden.

My use case is a manual movement of the Xfce panel (managed in the motion signal handler when the panel is clicked) until changing monitor. Then GTK Layer Shell remaps the panel (hide/show), and this "manual grab" is lost. It is inevitable that it is lost I think, but on Wayfire it happens properly, because the release event is received, while on Labwc this event is missing (so "grab" status is not reset, etc.).

**Comment by Consolatis** [comment]
> When a surface is clicked and is hidden before the button is released, the corresponding release event is never received.

Hmm.. interesting edge case indeed. We are storing the surface a mouse press started on to implement out-of-surface scrolling and to allow for movement between XDG popups (e.g. different menu popups) of the same toplevel.

So we could check for that stored surface in the unmap handler and send a fake release event, that should solve it.

**Comment by johanmalm** [comment]
> So we could check for that stored surface in the unmap handler and send a fake release event, that should solve it.

Thanks @Consolatis. Sounds sensible.

**Comment by Consolatis** [comment]
Did some testing with this example:
```python3
#!/usr/bin/env python3
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk

if __name__ == '__main__':
	def debug_click(source, event):
		if event.type == event.type.BUTTON_PRESS:
			print("Got press for button", event.button)
			source.hide()
		elif event.type == event.type.BUTTON_RELEASE:
			print("Got release for button", event.button)
			source.show()
		else:
			print("Got some unknown button event:", event.type)
		return True

	# button press / release test
	win = Gtk.Window()
	win.connect('button-press-event', debug_click)
	win.connect('button-release-event', debug_click)
	win.connect('destroy', Gtk.main_quit)
	win.show_all()

	Gtk.main()
```

And ran it via `WAYLAND_DEBUG=1 test_button_event.py`. Result:
```
[4079339.597] wl_pointer@7.button(196, 3079509632, 272, 1)
[4079339.685] wl_pointer@7.frame()
Got press for button 1
[4079340.649]  -> xdg_toplevel@23.destroy()
[4079340.707]  -> xdg_surface@22.destroy()
[4079340.736]  -> org_kde_kwin_server_decoration@24.release()
[4079340.765]  -> wl_surface@17.destroy()
[4079340.803]  -> wl_buffer@26.destroy()
[4079340.831]  -> wl_shm_pool@25.destroy()
[4079351.780] wl_display@1.delete_id(23)
[4079351.935] wl_display@1.delete_id(22)
[4079352.053] wl_display@1.delete_id(24)
[4079352.109] wl_display@1.delete_id(17)
[4079352.149] wl_display@1.delete_id(26)
[4079352.183] wl_display@1.delete_id(25)
[4079352.218] wl_pointer@7.leave(197, nil)
[4079352.255] wl_pointer@7.frame()
[4079352.290] wl_keyboard@3.leave(200, nil)
```

So GTK receives a `wl_pointer@7.leave(197, nil)` as it should. I am not so sure anymore if we as a compositor should fake a release event or if GTK should handle this situation instead. When it looses pointer focus while a button is pressed it could instead inject a release event itself. If we instead fake a release event it could potentially cause more issues like trigger a menu selection on a destroying submenu or similar (just a guess, not tested). Maybe the wayland pointer protocol gives some hints how to properly deal with this situation.

**Comment by notpeelz** [comment]
I just ran into this issue.

This causes the default behavior of my side mouse button (i.e. "navigate backward/previous") to trigger unexpectedly on some programs. Very annoying.

```xml
<context name="All">
  <mousebind button="Left" action="Press" />
  <mousebind button="Right" action="Press" />
  <mousebind button="Side" action="Press">
    <action name="Execute" command="*some command here*" />
  </mousebind>
</context>
```
